### ISSUE TITLE
S3: encryption header is missing when using a KMS key

### ISSUE THREAD

**Issue by irinarenteria** [mcohen01/amazonica#338]
I’ve been running into an issue where using a KMS key to encrypt a file in S3 returns a 400. This seems to be because S3 is expecting a different header than the one sent by Amazonica.

The error I’m seeing:

> com.amazonaws.services.s3.model.AmazonS3Exception: Server Side Encryption with AWS KMS managed key requires HTTP header x-amz-server-side-encryption : aws:kms (Service: Amazon S3; Status Code: 400; Error Code: InvalidArgument;

Steps to reproduce:

```clojure
(defcredential aws-access-key aws-secret-key aws-region-name)

(put-object :bucket-name "test-bucket"
            :key "test-bucket-key"
            :file "file"
            :metadata {:server-side-encryption-aws-kms-key-id "kms-key-ID"})
```

**Comment by irinarenteria** [comment]
@mcohen01 Assuming I've narrowed down the issue correctly, I've submitted a quick fix. Let me know your thoughts!

**Comment by mcohen01** [comment]
I'm not terribly familiar with this code but I think that error message is telling you that you have to also pass the `x-amz-server-side-encryption` header, with value "aws:kms." But I'm not sure. That's what the [AWS Docs](http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingKMSEncryption.html) seems to suggest. Can you try adding `(.setHeader om "x-amz-server-side-encryption" "aws:kms")` under the [`when-let`](https://github.com/mcohen01/amazonica/blob/master/src/amazonica/aws/s3.clj#L170) on your branch? If that works, can you confirm whether or not your bucket policy enforces this, as described in the doc linked to above? I ask that question because the [PR](https://github.com/mcohen01/amazonica/pull/243) that added support for KMS encryption did not include this header, so I'm guessing no one has ever actually used this functionality combined with bucket policies that enforce the KMS encryption. Finally, if you add the header on your branch and that works, can you toggle your bucket policy to *not* require encryption and confirm that passing the additional header still works? Thanks!!!

**Comment by irinarenteria** [comment]
Right on, @mcohen01, you're correct! That error message was indeed specifying that a missing header with the "aws:kms" value.

Curiously enough, this test bucket has no policy attached to it and no KMS restrictions as far as I can tell. I ran a `aws s3api get-bucket-policy --bucket test-bucket` check to make sure my eyes don't deceive me and nothing came up. Looks like it might be a "silent" requirement from AWS internals somewhere?

I added `(.setHeader om "x-amz-server-side-encryption" "aws:kms")` under that `when-let` branch and sending the object to S3 worked. I also added a policy to the bucket that requires that header and sent another KMS-encrypted file to the bucket and that worked too. So it seems like that fixes the issue!
